### PR TITLE
SAK-29373 add the ability to create auto groups from roster

### PR DESCRIPTION
**Pull request by bjones86**
https://jira.sakaiproject.org/browse/SAK-29373

Several faculty members have requested the ability to create Auto Groups from a selected roster, just like how they can create Auto Groups from a selected role.

The linked PR introduces the same set of choices you get for Auto Groups for a role for a selected roster as well:

    "Create a single group for the selected roster" (default)
    "Create random groups from members in the selected roster"

If you choose the later of the two options, you get two sub-options just like with role (random number of groups with specified number of users per group, or random number of users in a specified number of groups).

I continued with the existing design in such a way that you can effectively create groups (random or not) for both sections and rosters at the same time.

### CODE REVIEW

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 1]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
+                return -1;
+            }
+            else
+            {
+                try
+                {
+                    intToSplit = Integer.parseInt( numToSplit );
+                    if( intToSplit <= 0 )
+                    {
+                        if( isGroupSplit )
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: Would be cleaner if this section (and the one repeated below for NFE) was extracted into a method.

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 2]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
+        int intToSplit = validateAutoGroupsFields( optionAssign, groupSplit, numToSplitGroup, numToSplitUser, groupTitleGroup, groupTitleUser );
+        int rosterIntToSplit = validateAutoGroupsFields( rosterOptionAssign, rosterGroupSplit, rosterNumToSplitGroup, rosterNumToSplitUser, rosterGroupTitleGroup, rosterGroupTitleUser );
+        if( intToSplit == -1 || rosterIntToSplit == -1 )
+        {
+            return null;
+        }
 
-    		List<String> usersList = new ArrayList<String>();
+        List<String> rosterList = new ArrayList<>();
+        if (!selectedRosters.isEmpty())
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: shouldn't need this test for empty, the foreach loop should just skip an empty collection

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 3]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
+                }
+            }
+        }
 
-    		for(String role:roleList)
-    		{
-    			// get the authz group
-    			String siteReference = siteService.siteReference(site.getId());
+        List<String> roleList = new ArrayList<>();
+        if (!selectedRoles.isEmpty())
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: same as above comment

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 4]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
+     * This could be random number of users in specified number of groups,
+     * or vice versa.
+     * 
+     * @param rosters a list of provider IDs (should only ever be one in the list)
+     * @param unit the number of groups or users per group requested
+     */
+    private void createRandomGroupsForRoster( List<String> providerIDs, int unit )
+    {
+        String groupTitle = rosterGroupSplit ? rosterGroupTitleGroup: rosterGroupTitleUser;
+        if( !StringUtils.isEmpty( groupTitle ) )
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: try isNotBlank instead here, it will detect strings that consist only of whitespace

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 5]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
+                try
+                {
+                    Set<Membership> sectionMembers = cms.getSectionMemberships( providerID );
+                    for( Membership member : sectionMembers )
+                    {
+                        String userEID = member.getUserId();
+                        try
+                        {
+                            String userID = userDirectoryService.getUserId( userEID );
+                            if( !userList.contains( userID ) )
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: this could get expensive with larger sections, consider profiling with some alternative implementations like keeping the userList sorted and doing a binary instead of linear search, may or may not be an overall win

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 6]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
-    			Group group = site.addGroup();
-    			group.getProperties().addProperty(group.GROUP_PROP_WSETUP_CREATED, Boolean.TRUE.toString());
+        int groupCount = 0;
+        List<Group> groupList = new ArrayList<>();
+        while( groupCount < numOfGroups )
+        {
+            // Create the group
+            Group group = site.addGroup();
+            group.getProperties().addProperty( Group.GROUP_PROP_WSETUP_CREATED, Boolean.TRUE.toString() );
+            group.setTitle( groupTitle + "-" + (groupCount + 1) );
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: string concat in a loop

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 7]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
+    private void createRandomGroupsForRole( List<String> roleList, int unit )
+    {
+        String groupTitle = groupSplit ? groupTitleGroup : groupTitleUser;
+        if( !StringUtils.isEmpty( groupTitle ) )
+        {
+            List<String> usersList = new ArrayList<>();
+            for( String role : roleList )
+            {
+                // get the authz group
+                String siteReference = siteService.siteReference( site.getId() );
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: move this outside the loop, it won't change

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 8]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
-				Member member = site.getMember(userId);
-				group.addMember(userId, member.getRole().getId(), member.isActive(), false);
-				//remove this user now:
-				usersList.remove(index);
-    		}
-    	}
+                try
+                {
+                    // Build the user ID list
+                    AuthzGroup siteGroup = authzGroupService.getAuthzGroup( siteReference );
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: move outside loop

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 9]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
-    			}       		
+    /**
+     * Create random groups of users from the given role.
+     * This could be random number of users in specified number of groups,
+     * or vice versa.
+     * 
+     * @param roleList a list of roles (should only ever be one in the list)
+     * @param unit the number of groups or users per group requested
+     */
+    private void createRandomGroupsForRole( List<String> roleList, int unit )
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: if there should only ever be one role, change the parameter to take a single role instead of a list

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/rsf/GroupAutoCreateProducer.java [hunk 10]
```diff
@@ -118,6 +118,7 @@ public void fillComponents(UIContainer arg0, ViewParameters arg1, ComponentCheck
 		 
 		 // for the site rosters list
 		 List<String> siteRosters= handler.getSiteRosters(null);
+		 String[] optionValues = new String[] { Integer.toString( 1 ), Integer.toString( 2 ) };
```
Comment (plukasew) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/rsf/GroupAutoCreateProducer.java]: how about string literals here? { "1", "2" }

File: site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java [hunk 11]
```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
...
+    private void createRandomGroups( boolean isGroupSplit, List<String> userIDs, String groupTitle, int unit )
+    {
+        int numOfGroups = -1;
+        int numOfUsersPerGroup = -1;
+        int totalNumOfUsers = userIDs.size();
+        
+        // Determine number of users per group and number of groups to create
+        if( isGroupSplit )
+        {
+            numOfGroups = unit > totalNumOfUsers ? totalNumOfUsers : unit;
```
Comment (bbailla2) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: This seems to be where we're preventing the existence of more groups than there are users (same on line 1148). Arguably, we may want some feedback to the user in these situations.

Comment (bjones86) [inline site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java]: Remainders will be addressed in a future ticket. We have plans to provide another option in the future to determine how the user wishes to handle remainders (either add the remainders to a new group, or add them randomly to groups within the set being created).

**Comment by bjones86** [comment]
Refactored according to all comments provided.

**Comment by ottenhoff** [comment]
Nice job with code review and quick changes!  +1

**Comment by bjones86** [comment]
Thanks, @ottenhoff, @plukasew and @bbailla2 for your help with the review. I'll keep this open over night just in case some other comments roll in, otherwise I'll merge it into master in the morning.
